### PR TITLE
Various fixes for istio.io Ingress SDS test

### PR DESCRIPTION
**Pull request by nmittler**
This required several changes and new features:

- Support for multiple verifiers
- Support for custom output
- Separate streams of command output (stdout, stderr)
- Disallow duplicate snippet names
- Eliminate intermediate files when evaluating templates

### CODE REVIEW

**Comment by nmittler** [comment]
@howardjohn Is it possible to upgrade the version of curl we're using on the prow containers? The ingress SDS test is failing because I'm using the curl arg `--retry-connrefused`, which requires curl `7.52.0` or later.

**Comment by howardjohn** [comment]
I would think so, it's in the tools repo under docker/build-tools.
hopefully as simple as changing a version number somewhere

Is that version of curl really new though, or can we assume users will have
it?

**Comment by nmittler** [comment]
@howardjohn 

> Is that version of curl really new though, or can we assume users will have
it?

Looks like the latest curl version is `7.66.0`. `7.53.0` was released in [Feb 22, 2017](https://curl.haxx.se/changes.html#7_53_0).

File: pkg/test/istioio/script.go [hunk 1]
```diff
@@ -74,55 +97,171 @@ var _ Step = Script{}
...
+//
+//     # $snippet mysnippet syntax="bash"
+//     $ kubectl apply -f somefile.yaml
+//     # $endsnippet
+//
+//     # $snippet mysnippetoutput syntax="text"
+//     this is the actual output of the command
+//     # $endsnippet
+//
+// You can specify the which output stream of the command to use for the output with the `outputstream`
```
Comment (brian-avery) [inline pkg/test/istioio/script.go]: you can specify which output stream

File: tests/integration/istioio/README.md [hunk 2]
```diff
@@ -152,84 +152,8 @@ istioio.Script{
...
-The following verifiers are supported:
-
-- **token**: the default verifier, if not specified. Performs a token-based comparison of expected
-and actual output. The syntax supports the wildcard `?` character to skip comparison
-for a given token.
-- **contains**: verifies that the output contains the given string.
-- **notContains**: verifies that the output does not contain the given string.
-
-See the `istioio.Script` API for additional configuration options.
+See [istioio.Script](../../../pkg/test/istioio/script.go) for the available
```
Comment (brian-avery) [inline tests/integration/istioio/README.md]: I think it makes sense to continue describing the configuration options here. The docs format/bash shell commands pattern that we're following for test writing doesn't assume a lot of Go knowledge. I'm not sure that we want to tell writers to read the source to figure out how to use write tests.

File: pkg/test/util/curl/version.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+// RequireMinVersionOrFail calls RequireMinVersion. If it returns an error, fails the given test.
+func RequireMinVersionOrFail(t test.Failer, min *semver.Version) {
+	t.Helper()
+	if err := RequireMinVersion(min); err != nil {
+		t.Fatal(err)
+	}
+}
+
+// RequireMinVersion returns an error if the version of curl on the system meet the specified minimum.
```
Comment (brian-avery) [inline pkg/test/util/curl/version.go]: if the version of curl on the system does not meet the specified minimum.

**Comment by nmittler** [comment]
@howardjohn 

The error due to the fact that this command returns nothing in kind:

```bash
export INGRESS_HOST=$(kubectl -n istio-system \
get service istio-ingressgateway -o jsonpath='{.status.loadBalancer.ingress[0].ip}')
```

At the point this is called, we've deployed a gateway. Do you know if ingresses behave differently in kind? Perhaps I need to add a wait before getting this field?

**Comment by howardjohn** [comment]
there is no load balancer IP in kind, like with minikube. In the test framework we have some option like --istio.test.minikube (poorly named) that makes this work by not looking for the external IP and instead looks for a different IP (not exactly sure which one)

File: tests/integration/istioio/README.md (hunk 2 above)
Comment (nmittler) [reply tests/integration/istioio/README.md]: I'm a little torn on this. I would say that the docs *must* exist here in the source code for certain. The question is whether or not we duplicate some/all of this in the MD file. My concern is that this class is a bit in flux and it's going to be painful to keep this and the MD file in sync. If we're asking the docs team to write code anyway, I would think that reading comments on the library they're using would not be an incredibly high bar.

@geeknoid thoughts?

**Comment by nmittler** [comment]
@geeknoid @brian-avery PTAL

File: tests/integration/istioio/README.md (hunk 2 above)
Comment (geeknoid) [reply tests/integration/istioio/README.md]: Tests are written developers, it's fine to assume access to the source code and knowledge of Go.

**Pull request merged**